### PR TITLE
Flow tests with page objects

### LINKED ISSUE #34: Refactor flow tests to use page objects
Reference material:
 - https://github.com/appfolio/ae_page_objects
 - https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project/flow-tests-with-page-objects

The starting skeleton is on the `flowTestsWithPageObjects` branch.

### PR DESCRIPTION
**Pull request by wpliao1989**
Closes #34 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: app/views/images/index.html.erb [hunk 1]
```diff
@@ -1,12 +1,12 @@
-<%= render partial: 'filter', locals: { tags: @tags } %>
-<div class="row">
+<%= render partial: "filter", locals: { tags: @tags } %>
```
Comment (rmacklin) [inline app/views/images/index.html.erb]: What was the reason for making this change (and the similar one on line 9)?

File: test/flows/images_crud_test.rb [hunk 2]
```diff
@@ -7,34 +7,36 @@ class ImagesCrudTest < FlowTestCase
     new_image_page = images_index_page.add_new_image!
 
     tags = %w(foo bar)
-    new_image_page = new_image_page.create_image!(url: 'invalid', tags: tags.join(', ')).as_a(PageObjects::Images::NewPage)
-    assert_equal 'must be a valid URL', new_image_page.url.error_message
+    new_image_page = new_image_page.create_image!(image_url: 'invalid',
+                                                  tags: tags.join(', '))
+      .as_a(PageObjects::Images::NewPage)
```
Comment (rmacklin) [inline test/flows/images_crud_test.rb]: If you want to break this into multiple lines, the way I would suggest doing it is:
```ruby
new_image_page = new_image_page.create_image!(
  image_url: 'invalid',
  tags: tags.join(', ')
).as_a(PageObjects::Images::NewPage)
```
since it doesn't add a ton of unnecessary whitespace.

File: test/flows/images_crud_test.rb [hunk 3]
```diff
@@ -7,34 +7,36 @@ class ImagesCrudTest < FlowTestCase
...
+    assert_equal 'is invalid', new_image_page.url.error_message
 
     image_url = 'https://media3.giphy.com/media/EldfH1VJdbrwY/200.gif'
-    new_image_page.url.set(image_url)
+    image_show_page = new_image_page.create_image!(image_url: image_url, tags: tags.join(', '))
```
Comment (rmacklin) [inline test/flows/images_crud_test.rb]: Why did you change this from:
```ruby
new_image_page.url.set(image_url)
image_show_page = new_image_page.create_image!
```
?

File: test/flows/images_crud_test.rb [hunk 4]
```diff
@@ -7,34 +7,36 @@ class ImagesCrudTest < FlowTestCase
...
+    assert_equal 'New image created!', image_show_page.flash_message(:success)
 
-    image_show_page = new_image_page.create_image!
-    assert_equal 'You have successfully added an image.', image_show_page.flash_message(:success)
-
-    assert image_show_page.has_url?(image_url)
+    assert image_show_page.image?(image_url)
```
Comment (rmacklin) [inline test/flows/images_crud_test.rb]: Why did you rename the method to `image?`

File: test/flows/images_crud_test.rb [hunk 5]
```diff
@@ -7,34 +7,36 @@ class ImagesCrudTest < FlowTestCase
...
+    assert image_show_page.image_link?(image_url)
```
Comment (rmacklin) [inline test/flows/images_crud_test.rb]: This method name is also a bit confusing

File: test/flows/images_crud_test.rb [hunk 6]
```diff
@@ -7,34 +7,36 @@ class ImagesCrudTest < FlowTestCase
...
+    assert image_show_page.edit_image_link?
```
Comment (rmacklin) [inline test/flows/images_crud_test.rb]: This method does not have a descriptive name, so I had to look at the definition to figure out what it does. After looking at it, I don't think this method has any value. You can remove it and just use `assert_predicate image_show_page.edit_image_link, :present?`

File: test/flows/images_crud_test.rb [hunk 7]
```diff
@@ -7,34 +7,36 @@ class ImagesCrudTest < FlowTestCase
...
+    assert image_show_page.delete_image_link?
```
Comment (rmacklin) [inline test/flows/images_crud_test.rb]: Same for this one

File: test/flows/images_crud_test.rb [hunk 8]
```diff
@@ -7,34 +7,36 @@ class ImagesCrudTest < FlowTestCase
...
   test 'delete an image' do
     cute_puppy_url = 'http://ghk.h-cdn.co/assets/16/09/980x490/landscape-1457107485-gettyimages-512366437.jpg'
     ugly_cat_url = 'http://www.ugly-cat.com/ugly-cats/uglycat041.jpg'
-    Image.create!([
-      { url: cute_puppy_url, tag_list: 'puppy, cute' },
-      { url: ugly_cat_url, tag_list: 'cat, ugly' }
-    ])
+
+    Image.create!([{ url: cute_puppy_url, tag_list: 'puppy, cute' },
+                   { url: ugly_cat_url, tag_list: 'cat, ugly' }])
```
Comment (rmacklin) [inline test/flows/images_crud_test.rb]: Why did you make this change? I consider the unnecessary indentation caused by this change to be a regression

File: test/flows/images_crud_test.rb [hunk 9]
```diff
@@ -47,34 +49,73 @@ class ImagesCrudTest < FlowTestCase
...
     puppy_url_2 = 'http://ghk.h-cdn.co/assets/16/09/980x490/landscape-1457107485-gettyimages-512366437.jpg'
     cat_url = 'http://www.ugly-cat.com/ugly-cats/uglycat041.jpg'
-    Image.create!([
-      { url: puppy_url_1, tag_list: 'superman, cute' },
-      { url: puppy_url_2, tag_list: 'cute, puppy' },
-      { url: cat_url, tag_list: 'cat, ugly' }
-    ])
+    Image.create!([{ url: puppy_url_1, tag_list: 'superman, cute' },
+                   { url: puppy_url_2, tag_list: 'cute, puppy' },
+                   { url: cat_url, tag_list: 'cat, ugly' }])
```
Comment (rmacklin) [inline test/flows/images_crud_test.rb]: Same as https://github.com/AppFolioOnboarding/image-sharer-wpliao1989/pull/35#discussion_r74145798.

Even _if_ you wanted to make this formatting change (which I don't recommend), you should do so it in _its own commit_, because mixing formatting changes with functional changes bloats the diffs unnecessarily.

File: test/flows/images_crud_test.rb [hunk 10]
```diff
@@ -47,34 +49,73 @@ class ImagesCrudTest < FlowTestCase
...
+  test 'no image flash should not persist' do
+    images_index_page = PageObjects::Images::IndexPage.visit
+    assert_equal 'No images found', images_index_page.flash_message(:notice)
+
+    url = 'http://www.appfolio.com/images/html/apm-logo-v2.png'
+    tags = 'cats, dogs'
+    Image.create!(url: url, tag_list: tags)
+
+    images_index_page = PageObjects::Images::IndexPage.visit
+    assert_raises(AePageObjects::LoadingElementFailed) do
```
Comment (rmacklin) [inline test/flows/images_crud_test.rb]: It would be a better page object interface if your `flash_message` method handled this exception and returned `nil` when there is no flash message.

File: test/page_objects/document.rb [hunk 11]
```diff
@@ -1,7 +1,18 @@
...
+
+    FLASH_MSG = {
+      success: 'alert-success',
+      error: 'alert-error',
+      alert: 'alert-block',
+      notice: 'alert-info'
+    }.freeze
+
+    def flash_class_for(flash_type)
+      FLASH_MSG[flash_type.to_sym] || flash_type.to_s
```
Comment (rmacklin) [inline test/page_objects/document.rb]: Why did you add the `|| flash_type.to_s`?

File: test/page_objects/extensions/element_inline_error_message.rb [hunk 12]
```diff
@@ -6,7 +6,7 @@ def error_message
 
         Capybara.using_wait_time(0) do
           begin
-            parent.find('.help-block').text
+            parent.find('span.help-block').text
```
Comment (rmacklin) [inline test/page_objects/extensions/element_inline_error_message.rb]: Why did you make this change?

File: test/page_objects/images/show_page.rb [hunk 13]
```diff
@@ -2,29 +2,52 @@ module PageObjects
...
         yield node.driver.browser.switch_to.alert
       end
 
       def delete_and_confirm!
-        #TODO
+        delete_image_link.node.click
         alert = node.driver.browser.switch_to.alert
         alert.accept
         window.change_to(IndexPage)
       end
```
Comment (rmacklin) [inline test/page_objects/images/show_page.rb]: You can change this to
```ruby
def delete_and_confirm!
  delete(&:accept)
  window.change_to(IndexPage)
end
```

File: test/flows/image_flows_test.rb [hunk 14]
```diff
@@ -1,182 +0,0 @@
...
-    assert page.has_link?('dogs')
-
-    click_link('cats')
-    assert page.has_current_path?(images_path(tag: 'cats'))
-    assert page.has_selector?("a[href=\"#{image_path(image1)}\"]")
-    assert page.has_no_selector?("a[href=\"#{image_path(image2)}\"]")
-    assert page.has_selector?('.filter')
-  end
-
-  test 'click on tag from show' do
```
Comment (rmacklin) [inline test/flows/image_flows_test.rb]: Is this covered in the new tests?

File: test/flows/image_flows_test.rb [hunk 15]
```diff
@@ -1,182 +0,0 @@
...
-    assert page.has_no_selector?("a[href=\"#{image_path(image2)}\"]")
-    assert page.has_link?('Clear Filter')
-
-    click_link('Clear Filter')
-    assert page.has_current_path?(images_path)
-    assert page.has_selector?("a[href=\"#{image_path(image1)}\"]")
-    assert page.has_selector?("a[href=\"#{image_path(image2)}\"]")
-  end
-
-  test 'click on tag in tag index' do
```
Comment (rmacklin) [inline test/flows/image_flows_test.rb]: Is this covered in the new tests?

File: app/views/images/index.html.erb (hunk 1 above)
Comment (wpliao1989) [inline app/views/images/index.html.erb]: Hmm that was an error. They should be single quoted strings.

File: test/flows/images_crud_test.rb (hunk 2 above)
Comment (wpliao1989) [inline test/flows/images_crud_test.rb]: Good idea!

File: test/flows/images_crud_test.rb (hunk 3 above)
Comment (wpliao1989) [inline test/flows/images_crud_test.rb]: Yes I think this will be more clear that everything is done through `create_image!` method.

File: test/flows/images_crud_test.rb (hunk 4 above)
Comment (wpliao1989) [inline test/flows/images_crud_test.rb]: Every `has_` method was complained by rubocop and it suggested taking `has_` off.

File: test/flows/images_crud_test.rb (hunk 6 above)
Comment (wpliao1989) [inline test/flows/images_crud_test.rb]: Thanks! It's a lot better!

File: test/flows/images_crud_test.rb (hunk 9 above)
Comment (wpliao1989) [inline test/flows/images_crud_test.rb]: I was trying to fix rubocop errors. It was complaining about the first format.

File: test/flows/images_crud_test.rb (hunk 10 above)
Comment (wpliao1989) [inline test/flows/images_crud_test.rb]: Sounds good!

File: test/flows/images_crud_test.rb (hunk 3 above)
Comment (rmacklin) [inline test/flows/images_crud_test.rb]: I disagree. With your change, you're refilling the `tag_list` input, which means you're not testing that the value of that field was persisted from the previous form submission.

File: test/page_objects/document.rb (hunk 11 above)
Comment (wpliao1989) [inline test/page_objects/document.rb]: So it the `FLASH_MSG` hash returns nil, we'll have a default css class name.

File: test/flows/images_crud_test.rb (hunk 4 above)
Comment (rmacklin) [inline test/flows/images_crud_test.rb]: Right, but is `image?` a descriptive name? How about `showing_image?`?

File: test/page_objects/extensions/element_inline_error_message.rb (hunk 12 above)
Comment (wpliao1989) [inline test/page_objects/extensions/element_inline_error_message.rb]: My simple form is using `.help-block` for two things: the hint and the error message.
Hint is `div.help-bloack` and error is `span.help-block`. I should probably change the error class to its own class like `error`

File: test/flows/images_crud_test.rb (hunk 9 above)
Comment (rmacklin) [inline test/flows/images_crud_test.rb]: Ok. Did rubocop actually error on the previous style or was this just changed by an autoformat? If the former, then we need to configure the rubocop rules to accept the first - I was under the impression @bboe had already done that after we made the same change in apm_bundle. Rubocop enforces a lot of good changes, but this is not one of them, and we shouldn't blindly follow it because the individual cops are configurable.

File: test/page_objects/document.rb (hunk 11 above)
Comment (rmacklin) [inline test/page_objects/document.rb]: But that default class name would be whatever the caller passed in, which won't be found on the page. Our page _only_ supports the types listed in `FLASH_MSG`

File: test/page_objects/extensions/element_inline_error_message.rb (hunk 12 above)
Comment (rmacklin) [inline test/page_objects/extensions/element_inline_error_message.rb]: Yes that would be better. Simple form has a configuration option for that.

File: test/page_objects/document.rb (hunk 11 above)
Comment (wpliao1989) [inline test/page_objects/document.rb]: Yeah that's correct. Now I think removing `|| flash_type.to_s` is a better idea.

File: test/flows/images_crud_test.rb (hunk 9 above)
Comment (wpliao1989) [inline test/flows/images_crud_test.rb]: Sounds good. I'll look into it.

File: test/flows/images_crud_test.rb (hunk 3 above)
Comment (wpliao1989) [inline test/flows/images_crud_test.rb]: Hmm haven't thought about that case. Thanks for catching this!

File: test/flows/images_crud_test.rb (hunk 4 above)
Comment (wpliao1989) [inline test/flows/images_crud_test.rb]: I agree it's not very descriptive. I'll figure out a new name.

File: test/flows/image_flows_test.rb (hunk 15 above)
Comment (wpliao1989) [inline test/flows/image_flows_test.rb]: The above 2 tests are not converted but I think they are basically covered in `view images associated with a tag`. That test is saying: if the link looks like `xxx`, the index page should show `yyy`. So If the link `xxx` is shown on show and tag index page, it should also work.

So I think these 2 tests are just unnecessary overhead.

**Comment by wpliao1989** [comment]
Updated @rmacklin

File: test/page_objects/images/new_page.rb [hunk 16]
```diff
@@ -1,12 +1,21 @@
...
+
+      def create_image!(image_url: nil, tags: nil)
+        url.set image_url if image_url.present?
+        tag_list.set tags if tags.present?
+
+        node.click_button('Create Image')
 
-      def create_image!(url: nil, tags: nil)
-        #TODO
+        window.change_to(ShowPage, NewPage)
```
Comment (rmacklin) [inline test/page_objects/images/new_page.rb]: Typically this is done as
```ruby
window.change_to(ShowPage, self.class)
```
which indicates "this page is either going to change to the show page, or it'll rerender this page again". Additionally, if you were to rename `NewPage`, you wouldn't have to modify this line.

File: test/page_objects/images/edit_page.rb [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+      end
+
+      def tags
+        tag_list.node.value
+      end
+
+      def edit_image!(tags:)
+        tag_list.set tags
+        node.click_button('Update Image')
+        window.change_to(ShowPage)
```
Comment (rmacklin) [inline test/page_objects/images/edit_page.rb]: Is `ShowPage` the only page this can change to?

**Comment by rmacklin** [comment]
https://github.com/AppFolioOnboarding/image-sharer-wpliao1989/pull/35#discussion_r74144268 still applies

File: test/flows/images_crud_test.rb [hunk 18]
```diff
@@ -47,11 +52,38 @@ class ImagesCrudTest < FlowTestCase
...
+    image_show_page = image_to_update.view!
+    assert image_show_page.showing_image?(url)
+    assert image_show_page.showing_image_link?(url)
+    assert_predicate image_show_page.edit_image_link, :present?
+    assert_equal %w(apm logo), image_show_page.tags
+
+    image_edit_page = image_show_page.edit_image!
+    assert_equal tags, image_edit_page.tags
+
+    image_show_page = image_edit_page.edit_image!(tags: 'rockets, hello')
```
Comment (rmacklin) [inline test/flows/images_crud_test.rb]: In flow tests, it's a good idea to test submitting each form with invalid input. That's currently done in the test that fills in the new form, but not here in the test that fills in the edit form.

File: test/page_objects/images/new_page.rb (hunk 16 above)
Comment (wpliao1989) [inline test/page_objects/images/new_page.rb]: Got it.

**Comment by rmacklin** [comment]
Hmm, I see that you did the stories out of order, so the other two comments don't apply until https://github.com/AppFolioOnboarding/image-sharer-wpliao1989/issues/12 is done.

File: test/page_objects/images/edit_page.rb (hunk 17 above)
Comment (wpliao1989) [inline test/page_objects/images/edit_page.rb]: Yes, because the required tag validation is not merged yet.

File: test/flows/images_crud_test.rb (hunk 18 above)
Comment (wpliao1989) [inline test/flows/images_crud_test.rb]: Yeah I'll add that test in https://github.com/AppFolioOnboarding/image-sharer-wpliao1989/pull/29

**Comment by wpliao1989** [comment]
Updated @rmacklin

**Comment by rmacklin** [comment]
OK, go ahead and merge and then address those two comments in #29

**Comment by wpliao1989** [comment]
Sounds good!
